### PR TITLE
Fix circular import between accelerate.state and accelerate.utils

### LINKED ISSUE #4215: Second circular import: accelerate.state <-> accelerate.utils (same pattern as #4173)
While looking into #4173, I found a second circular import that follows the same pattern but doesn't go through `accelerate.utils.bnb`.

`accelerate/state.py` and `accelerate/optimizer.py` both import from the aggregating `accelerate.utils` package (`from .utils import ...`). Several submodules that `accelerate/utils/__init__.py` eagerly loads import back from `accelerate.state`:

```
src/accelerate/utils/operations.py:27:from ..state import AcceleratorState, PartialState
src/accelerate/utils/modeling.py:32:from ..state import AcceleratorState
src/accelerate/utils/other.py:32:from ..state import PartialState
src/accelerate/utils/tqdm.py:22:from ..state import PartialState
src/accelerate/utils/random.py:21:from ..state import AcceleratorState
```

So `accelerate.state` depends on `accelerate.utils`, and `accelerate.utils` depends back on `accelerate.state`, the same shape as the `utils <-> big_modeling` cycle in #4173.

Single threaded imports work fine today because `accelerate/__init__.py` imports `.accelerator` first, and that chain happens to finish loading the full `utils` package before `state.py` gets imported, so the cycle never actually gets exercised. That's the same "ordering accident" described in #4173, just at a different pair of entry points. The reproducer from that issue triggers it here too, with `accelerate.state` and `accelerate.utils.operations` as the two names raced from separate threads.

I've opened a fix that has `state.py` and `optimizer.py` import from the specific leaf modules that define what they need (`utils/dataclasses.py`, `utils/environment.py`, `utils/imports.py`, `utils/operations.py`) instead of the aggregating package, the same remediation already used for `bnb.py`.

One thing worth flagging separately: `accelerate/__init__.py` imports `.accelerator` before anything else, and `accelerator.py` itself pulls in most of the rest of the package. Because of that, concurrent first imports of two different `accelerate` submodules will contend on that same shared bootstrap regardless of which specific internal cycle gets fixed, and CPython's import lock can hit its deadlock detector even without a real cycle involved once initialization takes any noticeable time. I don't think that's something a small PR can reasonably fix, since it would mean rethinking how much `accelerate/__init__.py` and `accelerator.py` load eagerly, but it seems worth a maintainer discussion since it means this bug class is likely to keep resurfacing at new entry points as more submodules get added or reordered.

### PR DESCRIPTION
**Pull request by Rakshit-gen**
Fixes #4215

`accelerate/state.py` and `accelerate/optimizer.py` imported from the aggregating `accelerate.utils` package. A handful of submodules that `accelerate/utils/__init__.py` eagerly loads (`operations.py`, `modeling.py`, `other.py`, `tqdm.py`, `random.py`) import `AcceleratorState` or `PartialState` back from `accelerate.state`, so `state.py` depending on the whole `utils` package created a real circular import, the same shape as the `utils <-> big_modeling` cycle from #4173.

It resolves fine in a normal single threaded import because `accelerate/__init__.py` always finishes loading `accelerator.py`, and with it the full `utils` package, before it reaches `state.py`. That ordering stops protecting you once two threads import different `accelerate` submodules at the same time.

## Changes

- `state.py` now imports `DistributedType`, `DynamoBackend`, `GradientAccumulationPlugin` and `SageMakerDistributedType` from `utils.dataclasses`, the environment helpers from `utils.environment`, and the availability checks from `utils.imports`, instead of the aggregating `utils` package.
- `optimizer.py` now imports `DistributedType` from `utils.dataclasses`, `is_lomo_available`/`is_torch_xla_available` from `utils.imports`, and `honor_type` from `utils.operations`, instead of the aggregating `utils` package.

Both are self-contained leaf modules with no import back to `accelerate.state` or `accelerate.optimizer`, so this removes the cycle without changing what either file actually uses.

## Tests

Added `CircularImportTester` to `tests/test_imports.py`. It statically parses `state.py` and `optimizer.py` and asserts neither one imports the `utils` package directly, so a future PR reintroducing `from .utils import ...` in either file will fail this test. I checked it against the pre-fix source and it fails there, and passes after this change.

Also ran the existing `tests/test_optimizer.py` and `tests/test_scheduler.py` suites and a CPU-only smoke test that prepares a model, optimizer and scheduler through `Accelerator` and runs a training step, both pass.

### CODE REVIEW

**Comment by Rakshit-gen** [comment]
@SunMarc could you take a look when you get a chance?

**Comment by SunMarc** [comment]
not a real issue, so closing.

**Comment by Rakshit-gen** [comment]
@SunMarc can you explain why?
According to me the issue is real because state.py imports from utils, while some utils modules import back from state.py, creating a circular dependency.

It works in normal imports because the current import order happens to hide the problem. With concurrent imports, that ordering isn't guaranteed and it can fail with a partially initialized module.

The fix removes this dependency entirely by importing directly from the specific utils modules. I can also add a minimal reproducer if needed.
